### PR TITLE
refactor(v2): conn manager, dbproxy, hoststore, rolestore, vault, nam, userstore

### PR DESCRIPTION
**Pull request by iljaSL**

### CODE REVIEW

File: api/rolestore/client.go [hunk 1]
```diff
@@ -19,734 +20,732 @@ type RoleStore struct {
...
+	return keys, err
 }
 
-// ImportPrincipalKey import new principal key for existing role
-func (store *RoleStore) ImportPrincipalKey(key PrivateKey, roleID string) (string, error) {
-	var object struct {
-		ID string `json:"id"`
-	}
+// CreateAuthorizedKey create authorized key for user.
+func (c *RoleStore) CreateAuthorizedKey(userID string, key *AuthorizedKey) (response.Identifier, error) {
```
Comment (cuongssh) [inline api/rolestore/client.go:318]: CreateUserAuthorizedKey

File: api/rolestore/client.go [hunk 2]
```diff
@@ -19,734 +20,732 @@ type RoleStore struct {
...
 
-	return object.ID, err
+	return identifier, err
 }
 
-// PrincipalKey returns a role's principal key object.
-func (store *RoleStore) PrincipalKey(roleID, keyID string) (*PrincipalKey, error) {
-	key := &PrincipalKey{}
+// GetUsersAuthorizedKey get users authorized key by id.
+func (c *RoleStore) GetUsersAuthorizedKey(userID, keyID string) (*AuthorizedKey, error) {
```
Comment (cuongssh) [inline api/rolestore/client.go:329]: GetUser (not plural)

File: api/rolestore/client.go [hunk 3]
```diff
@@ -19,734 +20,732 @@ type RoleStore struct {
...
-	return result.Items, err
+	return users, err
 }
 
-// GeneratePrincipalKey generate new principal key for existing role
-func (store *RoleStore) GeneratePrincipalKey(roleID string) (string, error) {
-	var object struct {
-		ID string `json:"id"`
+// GetUsersAuthorizedKeys get users authorized keys.
+func (c *RoleStore) GetUsersAuthorizedKeys(userID string, opts ...filters.Option) (*response.ResultSet[AuthorizedKey], error) {
```
Comment (cuongssh) [inline api/rolestore/client.go:301]: GetUser (not plural)

File: api/rolestore/client.go [hunk 4]
```diff
@@ -19,734 +20,732 @@ type RoleStore struct {
...
 
 	return key, err
 }
 
-// DeletePrincipalKey delete a role's principal key
-func (store *RoleStore) DeletePrincipalKey(roleID, keyID string) error {
-	_, err := store.api.
-		URL("/role-store/api/v1/roles/%s/principalkeys/%s", roleID, keyID).
+// UpdateUsersAuthorizedKey update users authorized key.
+func (c *RoleStore) UpdateUsersAuthorizedKey(userID, keyID string, key *AuthorizedKey) error {
```
Comment (cuongssh) [inline api/rolestore/client.go:340]: UpdateUser (not plural)

File: api/rolestore/client.go [hunk 5]
```diff
@@ -19,734 +20,732 @@ type RoleStore struct {
...
+	_, err := c.api.
+		URL("/role-store/api/v1/users/%s/authorizedkeys/%s", userID, keyID).
+		Put(&key)
+
+	return err
+}
+
+// DeleteUsersAuthorizedKey delete a users authorized key.
+func (c *RoleStore) DeleteUsersAuthorizedKey(userID, keyID string) error {
```
Comment (cuongssh) [inline api/rolestore/client.go:349]: DeleteUser...

File: api/rolestore/client.go [hunk 6]
```diff
@@ -19,734 +20,732 @@ type RoleStore struct {
...
 		Delete()
 
 	return err
 }
 
-// User gets information about the argument user ID.
-func (store *RoleStore) User(userID string) (*User, error) {
-	user := &User{}
+// GetCurrentUsersAuthorizedKeys get current users authorized keys.
+func (c *RoleStore) GetCurrentUsersAuthorizedKeys(opts ...filters.Option) (*response.ResultSet[AuthorizedKey], error) {
```
Comment (cuongssh) [inline api/rolestore/client.go:358]: For all the functions below with /current/ in path: decide which pattern to name:
- <Get>CurrentUser<Stuff>, or
- <Get><Stuff>CurrentUser

Apply the same pattern to all eligible functions after this line.

File: api/rolestore/model.go [hunk 7]
```diff
@@ -7,339 +7,483 @@
...
+	MobileMfaStatus string  `json:"mobile_mfa_status"`
+}
+
+type MFASeed struct {
+	Seed_string  string `json:"seed_string,omitempty"`
+	Seed_qr_code string `json:"seed_qr_code,omitempty"`
+}
+
+// Attributes user attributes definition.
+type Attributes struct {
```
Comment (cuongssh) [inline api/rolestore/model.go:284]: UserAttribute (also, not plural)

File: api/rolestore/model.go [hunk 8]
```diff
@@ -7,339 +7,483 @@
...
-	Sortdir   string `json:"sortdir,omitempty"`
-	Sortkey   string `json:"sortkey,omitempty"`
-	Tokencode string `json:"tokencode,omitempty"`
-	Refresh   bool   `json:"refresh,omitempty"`
-	Offset    int    `json:"offset,omitempty"`
-	Limit     int    `json:"limit,omitempty"`
-	TTL       int    `json:"ttl,omitempty"`
+// AWSRoleParams aws role query parameter definition.
+type AWSRoleParams struct {
+	Refresh string `url:"refresh,omitempty"`
```
Comment (cuongssh) [inline api/rolestore/model.go:17]: Should be bool, same as API spec.

File: api/connectionmanager/client.go [hunk 9]
```diff
@@ -19,403 +19,393 @@ type ConnectionManager struct {
...
 
 	return err
 }
 
-// StartAnalyzing start ueba analysis
-func (store *ConnectionManager) StartAnalyzing(datasetID string) error {
-	_, err := store.api.
-		URL("/connection-manager/api/v1/ueba/start-analyzing/%s", url.PathEscape(datasetID)).
+// StartAnalyzing start analyzing connections with a saved dataset.
+func (c *ConnectionManager) StartAnalyzing(datasetID string) error {
```
Comment (cuongssh) [inline api/connectionmanager/client.go:274]: StartUebaAnalyzing, same for Stop...

File: api/connectionmanager/client.go [hunk 10]
```diff
@@ -19,403 +19,393 @@ type ConnectionManager struct {
...
 		Post(nil, &count)
 
 	return count, err
 }
 
-// ConnectionCounts Get number of connections for dataset with given parameters.
-// All connections, if json empty in body.
-func (store *ConnectionManager) ConnectionCounts(timerange TimeRange) (ConnectionCount, error) {
+// GetConnectionCounts get number of connections for dataset.
+func (c *ConnectionManager) GetConnectionCounts(timeRange TimeRange) (ConnectionCount, error) {
```
Comment (cuongssh) [inline api/connectionmanager/client.go:361]: Put Ueba to all function names that deal with Ueba

File: api/secretsmanager/client.go [hunk 11]
```diff
@@ -3,452 +3,464 @@ package secretsmanager
...
-func (s *Client) TargetDomains(offset, limit int, sortkey, sortdir string) (TdResult, error) {
-	result := TdResult{}
-	filters := Params{
-		Offset:  offset,
-		Limit:   limit,
-		Sortkey: sortkey,
-		Sortdir: sortdir,
+// MARK: Manage Secrets
+// GetHostSecretMetadata get host secret metadata for all accounts.
+func (c *SecretsManager) GetHostSecretMetadata(hostID string) (*PostHostSecret, error) {
```
Comment (cuongssh) [inline api/secretsmanager/client.go:160]: PostHostSecret sounds strange as a type

File: api/secretsmanager/model.go [hunk 12]
```diff
@@ -2,13 +2,46 @@ package secretsmanager
...
+type PostHostSecret struct {
+	Metadata HostSecret      `json:"metadata" validate:"required"`
+	Accounts []AccountSecret `json:"accounts" validate:"required"`
 }
 
+// AccountSecret account secret definition.
+type AccountSecret struct {
+	Account           string     `json:"account"`
+	HostID            string     `json:"host_id"`
+	RotationInitiated *time.Time `json:"rotation_initiated"`
```
Comment (cuongssh) [inline api/secretsmanager/model.go:15]: All pointers usually come with omitempty

File: api/userstore/model.go [hunk 13]
```diff
@@ -8,115 +8,110 @@ package userstore
...
-	AuthClientSecret string              `json:"oauth_client_secret"`
-	Roles            []rolestore.RoleRef `json:"roles,omitempty"`
-	Created          string              `json:"created,omitempty"`
-	Author           string              `json:"author,omitempty"`
+// APIClientSearch api client search request definition.
+type APIClientSearch struct {
+	Keywords string `json:"keywords"`
+	SortDir  string `json:"sortdir"`
+	SortKey  string `json:"sortkey"`
+	Limit    *int   `json:"limit"`
```
Comment (cuongssh) [inline api/userstore/model.go:76]: omitempty for pointer type

**Review by cuongssh — CHANGES_REQUESTED** [review]

File: api/rolestore/client.go [hunk 14]
```diff
@@ -19,734 +20,732 @@ type RoleStore struct {
...
 
 	return err
 }
 
-// UpdateAWSRoleLink update existing aws role
-func (store *RoleStore) UpdateAWSRoleLink(awsRoleID string, roles []RoleRef) error {
-	_, err := store.api.
-		URL("/role-store/api/v1/awsroles/%s/roles", url.PathEscape(awsRoleID)).
+// GetLinkedRoles get AWS role granting PrivX roles.
+func (c *RoleStore) GetLinkedRoles(awsRoleID string) (*response.ResultSet[AWSRole], error) {
```
[the remark this answers is not in the document]
Comment (iljaSL) [reply api/rolestore/client.go:140]: Good catch, LinkedPrivXRole should be returned instead.

File: api/rolestore/model.go (hunk 7 above)
Comment (iljaSL) [reply api/rolestore/model.go:284]: That has been taken from the original struct definition, but let's name it differently in the sdk, UserAttribute makes more sense in this case.

File: api/secretsmanager/client.go (hunk 11 above)
Comment (iljaSL) [reply api/secretsmanager/client.go:160]: Yep it does, this is taken from the original struct definition. Same goes also for the create handler which takes PostHostSecret and returns a PostHostSecret...
Let me remove the "Post" suffix for this struct, this should make it a bit less confusing.

File: api/rolestore/client.go [hunk 15]
```diff
@@ -19,734 +20,732 @@ type RoleStore struct {
...
 
-	_, err := store.api.
-		URL("/role-store/api/v1/roles/resolve").
-		Post(&names, &result)
+	return err
+}
 
-	return result.Items, err
+// SetMFA enable, disable or reset mfa authentication.
+func (c *RoleStore) SetMFA(userIDs []string, action string) error {
```
[the remark this answers is not in the document]
Comment (iljaSL) [reply api/rolestore/client.go]: I included a enum type to the models.

**Review by cuongssh — APPROVED** [review]

**Pull request merged**